### ISSUE TITLE
Optimization to reorder commutative computations

### ISSUE THREAD

**Issue by mrocklin** [ContinuumIO/dask#79]
Some function pairs can commute to achieve the same result but with less cost.  It would be nice to optimize the graph to commute these functions accordingly.

E.g. if we have `(f, (g, x))` and know that this is equivalent-but-slower-than `(g, (f, x))` then we might want to make that switch for all such `x` in the graph.

This is very important when interacting with column-stores like BColz.  It's very nice to change from 

    (getitem, (convert_to_dataframe, my_column_store), 'my-column')

to 

    (convert_to_dataframe, (getitem, my_column_store, 'my-column'))

This often confers an order of magnitude reduction of I/O.  cc @eriknw

**Comment by jcrist** [comment]
This isn't *really* commutativity, as the functions, not the arguments are commuting. As such, syntactic matching rewrite rules should easily be able to handle this. The big question is how to determine if an operation is more or less costly, based on knowledge of the functions and the arguments.

**Comment by mrocklin** [comment]
Valid point.  Yes, syntactic matching rewrite rules would be sufficient.  It might be an interesting challenge to write this with sufficient efficiency not to bog down overhead in the common case.  Although the fact that we're using straight tuples should make this much faster than it would be in other projects like SymPy.

I'm happy to supply function-pairs/patterns that I always want to swap.

**Comment by jcrist** [comment]
The never-ending pattern matching project I've been working on should be able to handle this with little modification. If you can supply a couple rewrite rules for testing, I should be able to work something up.

**Comment by mrocklin** [comment]
```Python
In [1]: f, g = 'fg'

In [2]: {('x', i): (g, (f, ('data', i)), 'column') for i in range(5)}  # Before
Out[2]: 
{('x', 0): ('g', ('f', ('data', 0)), 'column'),
 ('x', 1): ('g', ('f', ('data', 1)), 'column'),
 ('x', 2): ('g', ('f', ('data', 2)), 'column'),
 ('x', 3): ('g', ('f', ('data', 3)), 'column'),
 ('x', 4): ('g', ('f', ('data', 4)), 'column')}

In [3]: {('x', i): (f, (g, ('data', i), 'column')) for i in range(5)}  # Desired result
Out[3]: 
{('x', 0): ('f', ('g', ('data', 0), 'column')),
 ('x', 1): ('f', ('g', ('data', 1), 'column')),
 ('x', 2): ('f', ('g', ('data', 2), 'column')),
 ('x', 3): ('f', ('g', ('data', 3), 'column')),
 ('x', 4): ('f', ('g', ('data', 4), 'column'))}
```

Not sure how I should spell the rules, but how about

    (g, (f, x), y) -> (f, (g, x, y))  for any x, y

**Comment by jcrist** [comment]
So, I have a still buggy, but functioning version of this. However, I just realized this may not be satisfactory, because of variable args functions.

Syntactic matching assumes that all functions have fixed arity (i.e. 'f' always has 1 arg, 'g' always has 2). However, this isn't always true in Python. I can think of some ways to fix this, but it'll require some work. Is this necessary (I think it probably is...)?

Also, is there need for conditional rewriting? As dasks can have strings instead of actual values, I'm not sure how much introspection into the arguments there could be.

**Comment by mrocklin** [comment]
Operationally the rules we need to apply are relatively simple and few in number.  We're in a much simpler regime than computer algebra.  I'm more concerned about performance (can we keep down the cost to sub-millisecond per task?) rather than rule complexity.

The rule that I described above is valuable and has neither of the concerns that you raise.

**Comment by jcrist** [comment]
Excellent. I'm able to apply a set of rules to a dask in the microsecond range on my fairly old laptop. I'll push a WIP PR shortly.

**Comment by jcrist** [comment]
Can this be closed? Did #82 fully resolve the issue?

**Comment by mrocklin** [comment]
Yup.  Closed.  Thanks again for syntactic rewriting.   It's turning out to be super-useful.
